File: Leetcode Challenge/January/Stone Game IV.cpp
class Solution {
public:
              
    int dp[100001];
    
    int helper(int n)
    {
        if(n<=0) return 0;
      if(dp[n]!=-1) return dp[n];    
        for(int i=1;i*i<=n;i++)
        {
            // if 2nd player return false then first will win; 
            if(helper(n-i*i)==0) return dp[n]=1;
        }
        
        return dp[n]=false;
        
    }
    bool winnerSquareGame(int n) {
        cout<<sizeof(dp);
        memset(dp,-1,sizeof(dp));
        return helper(n);
    }
};

File: Leetcode Challenge/January/Java/Stone Game IV.java
class Solution {
    Boolean[] dp = new Boolean[100001];

        public boolean winnerSquareGame(int n) {
            if (n == 0) return false;
            if (dp[n] != null) return dp[n];
            boolean winner = false;
            for (int i = 1; i * i <= n; i++) {
                if (!winnerSquareGame(n - i * i)) {
                    winner = true;
                    break;

                }
            }
            return dp[n] = winner;
        }
}
